### PR TITLE
#37 Sorting Topics/Lessons

### PR DESCRIPTION
**Pull request by qunabu**
Mentions #37

### CODE REVIEW

File: src/Repositories/TopicRepository.php [hunk 1]
```diff
@@ -161,6 +161,14 @@ public function update(array $input, int $id): Topic
 
             //$model->topicable()->delete();
             $model->topicable()->associate($content['model'])->save();
+        } elseif (isset($input['topicable_type']) && $model->topicable_type == $input['topicable_type']) {
+            $classType = $input['topicable_type'];
+
+            if (method_exists($classType, 'createResourseFromRequest')) {
```
Comment (pa-cholek) [inline src/Repositories/TopicRepository.php:167]: `Resource` and not `Resourse` (typo)

File: src/Http/Requests/SortAPIRequest.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+        if (isset($user) && $user->can('update', $course)) {
+            $class = $this->input('class');
+            $ids = array_map(function ($order) {
+                return $order[0];
+            }, $this->input('orders'));
+
+            switch ($class) {
+                case "Lesson":
+                    $lessons = Lesson::whereIn('id', $ids)->get();
```
Comment (pa-cholek) [inline src/Http/Requests/SortAPIRequest.php:30]: Couldn't this query and loop be simplified to
```suggestion
           return Lesson::whereIn('id', $ids)->where('course_id', '!=', $course->id)->doesntExist();
```

File: src/Http/Requests/SortAPIRequest.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                    foreach ($lessons as $lesson) {
+                        if ($lesson->course_id !== $course->id) {
+                            return false; // id from array is not matching course
+                        }
+                    }
+                    break;
+
+                case "Topic":
+                    $topics = Topic::with('lesson')->whereIn('id', $ids)->get();
```
Comment (pa-cholek) [inline src/Http/Requests/SortAPIRequest.php:39]: Maybe something like
```suggestion
                  return Topic::whereIn('id', $ids)->whereHave('lesson', function (Builder $query) {
    $query->where('course_id', $course->id);
})->doesntExist();
```

**Review by pa-cholek — COMMENTED** [review]
I'm not sure if there shouldn't be two separate Request classes and `class` input param moved to be part of route. Will there be more possible values of `class` param in future?

**Comment by qunabu** [comment]
> I'm not sure if there shouldn't be two separate Request classes and `class` input param moved to be part of route. Will there be more possible values of `class` param in future?

that was my point to be easy extendable

**Pull request merged**